### PR TITLE
add BGL and basis handling in moments

### PR DESCRIPTION
**Pull request by MiroDudik**
This PR contains an initial implementation of BGL and also introduces the basis generation support within moments.

Basis generation is required to integrate moments and GridSearch. The current implementation should be enough to get started.

### CODE REVIEW

File: fairlearn/reductions/moments/bounded_group_loss.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        zero_vec = pd.Series(0.0, self.index)
+        i = 0
+        for attr in attr_vals:
+            self.pos_basis[i] = 0 + zero_vec
+            self.neg_basis[i] = 0 + zero_vec
+            self.pos_basis[i][attr] = 1
+            self.neg_basis_present.at[i] = False
+            i += 1
+
+    def gamma(self, predictor):
```
Comment (riedgar-ms) [inline fairlearn/reductions/moments/bounded_group_loss.py]: Should we make gamma a 'virtual' method (for Python purposes, throw NotImplemented) in the base Moment class?

File: fairlearn/reductions/moments/bounded_group_loss.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        self.tags["pred"] = predictor(self.X)
+        self.tags["loss"] = self.reduction_loss.eval(self.tags["label"], self.tags["pred"])
+        expect_attr = self.tags.groupby("protected_attribute").mean()
+        self._gamma_descr = str(expect_attr[["loss"]])
+        return expect_attr["loss"]
+
+    def project_lambda(self, lambda_vec):
+        return lambda_vec
+
+    def signed_weights(self, lambda_vec):
```
Comment (riedgar-ms) [inline fairlearn/reductions/moments/bounded_group_loss.py]: Similarly, should this be an 'abstract' method in `Moment` ?

File: fairlearn/reductions/moments/bounded_group_loss.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        super().__init__(loss, no_groups=True)
+
+
+class GroupLossMoment(LossMoment):
+
+    def __init__(self, loss):
+        super().__init__(loss, no_groups=False)
+
+
+class SquareLoss:
```
Comment (riedgar-ms) [inline fairlearn/reductions/moments/bounded_group_loss.py]: Does this actually belong under metrics? Similar for the other losses you define?

Comment (MiroDudik) [reply fairlearn/reductions/moments/bounded_group_loss.py]: these differ from metrics in two ways: (1) they are required to be bounded (and that's recorded in min/max fields), and (2) they operate on vectors componentwise. But happy to consider moving them elsewhere.

File: fairlearn/reductions/moments/bounded_group_loss.py (hunk 2 above)
Comment (MiroDudik) [reply fairlearn/reductions/moments/bounded_group_loss.py]: Yep. I think we should move all of the "required" functionality into Moment.

File: fairlearn/reductions/moments/bounded_group_loss.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        self.reduction_type = _REDUCTION_TYPE_LOSS_MINIMIZATION
+        self.reduction_loss = loss
+        self.no_groups = no_groups
+
+    def default_objective(self):
+        return AverageLossMoment(self.reduction_loss)
+
+    def init(self, dataX, dataA, dataY):
+        if self.no_groups:
+            dataA = dataY.apply(lambda y: "all")
```
Comment (riedgar-ms) [inline fairlearn/reductions/moments/bounded_group_loss.py]: For my edification.... does this correspond to the 'event' $\Epsilon_{\star}$ in the papers?

File: fairlearn/reductions/moments/bounded_group_loss.py (hunk 3 above)
Comment (riedgar-ms) [reply fairlearn/reductions/moments/bounded_group_loss.py]: We can leave them here for now. Especially with my other modifications to metrics still in PR

File: fairlearn/reductions/moments/bounded_group_loss.py (hunk 4 above)
Comment (MiroDudik) [reply fairlearn/reductions/moments/bounded_group_loss.py]: yes :-)... once we move to load_data(self, X, y, **kwargs), I expect this to be handled more transparently.

File: fairlearn/reductions/moments/bounded_group_loss.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import pandas as pd
+import numpy as np
+from .moment import Moment
+from .moment import _REDUCTION_TYPE_LOSS_MINIMIZATION
+
+
+class LossMoment(Moment):
```
Comment (riedgar-ms) [inline fairlearn/reductions/moments/bounded_group_loss.py]: I have also been meaning to ask.... can we write useful unit tests of these, with carefully chosen X, Y and A?

File: fairlearn/reductions/moments/bounded_group_loss.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        self.no_groups = no_groups
+
+    def default_objective(self):
+        return AverageLossMoment(self.reduction_loss)
+
+    def init(self, dataX, dataA, dataY):
+        if self.no_groups:
+            dataA = dataY.apply(lambda y: "all")
+        super().init(dataX, dataA, dataY)
+        self.prob_attr = self.tags.groupby("protected_attribute").size() / self.n
```
Comment (romanlutz) [inline fairlearn/reductions/moments/bounded_group_loss.py]: would be nice if "protected_attribute" is defined as a constant :-) we can do that later, too, if you don't want to do that now

**Comment by romanlutz** [comment]
Still needs tests and documentation, but we can open an issue for this.

File: fairlearn/reductions/moments/conditional_opportunity.py [hunk 7]
```diff
@@ -18,6 +27,24 @@ def init(self, dataX, dataA, dataY, dataGrp):
                            keys=["+", "-"],
                            names=["sign", "grp", "protected_attribute"])
         self.index = signed.index
+        self.default_objective_lambda_vec = None
+
+        # fill in the information about the basis
+        grp_vals = self.tags["grp"].unique()
```
Comment (romanlutz) [inline fairlearn/reductions/moments/conditional_opportunity.py]: I really got confused with the naming when I started working on this, so maybe I'll open an issue to rename
grp -> group
attr -> attribute
vec -> vector (although that's the least unclear)
pos -> positive
neg -> negative

File: fairlearn/reductions/moments/bounded_group_loss.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        self.max_val = max_val
+        self.min = 0
+        self.max = (max_val-min_val) ** 2
+
+    def eval(self, y_true, y_pred):
+        return (np.clip(y_true, self.min_val, self.max_val)
+                - np.clip(y_pred, self.min_val, self.max_val)) ** 2
+
+
+class AbsLoss:
```
Comment (romanlutz) [inline fairlearn/reductions/moments/bounded_group_loss.py]: is this "absolute"? Can we spell it out?

Comment (MiroDudik) [reply fairlearn/reductions/moments/bounded_group_loss.py]: sure.

File: fairlearn/reductions/moments/conditional_opportunity.py (hunk 7 above)
Comment (MiroDudik) [reply fairlearn/reductions/moments/conditional_opportunity.py]: Yes. conditional_opportunity has a variety of naming issues (beginning with the name "conditional_opportunity"). I'd like to stick to "vec" to keep things concise.

**Comment by MiroDudik** [comment]
The key comments are: renaming of variables and unit tests. For the sake of expediency, I'm going to merge this and address some of the renaming comments in the next PR.

**Pull request merged**